packages/gatsby-plugin-mdx/src/gatsby-node.ts

```diff
@@ -1,53 +1,90 @@
 import type { GatsbyNode, NodeInput } from "gatsby"
 import type { FileSystemNode } from "gatsby-source-filesystem"
-import type { Options } from "@mdx-js/loader"
+import type { IFileNode, NodeMap } from "./types"
 
 import path from "path"
 import { sentenceCase } from "change-case"
 import fs from "fs-extra"
 import grayMatter from "gray-matter"
 
-import { defaultOptions, IMdxPluginOptions } from "./plugin-options"
-import { IMdxLoaderOptions } from "./mdx-loader"
+import {
+  defaultOptions,
+  enhanceMdxOptions,
+  IMdxPluginOptions,
+} from "./plugin-options"
+import { IGatsbyLayoutLoaderOptions } from "./gatsby-layout-loader"
+import compileMDX from "./compile-mdx"
+import { IGatsbyMDXLoaderOptions } from "./gatsby-mdx-loader"
 
 /**
  * Add support for MDX files including using Gatsby layout components
  */
-export const onCreateWebpackConfig: GatsbyNode["onCreateWebpackConfig"] = (
-  { actions, loaders, getNodesByType },
-  pluginOptions
-) => {
-  const fileNodes = getNodesByType(`File`)
-  const fileMap = new Map()
-  fileNodes.forEach(fileNode => fileMap.set(fileNode.absolutePath, fileNode))
-
-  actions.setWebpackConfig({
-    module: {
-      rules: [
-        {
-          test: /\.mdx?$/,
-          use: [
-            loaders.js(),
-            {
-              loader: `@mdx-js/loader`,
-              options: {
-                useDynamicImport: true,
-                providerImportSource: `@mdx-js/react`,
-              } as Options,
-            },
-            {
-              loader: path.join(`gatsby-plugin-mdx`, `dist`, `mdx-loader`),
-              options: {
-                fileMap,
-                pluginOptions,
-              } as IMdxLoaderOptions,
-            },
-          ],
-        },
-      ],
-    },
-  })
-}
+export const onCreateWebpackConfig: GatsbyNode["onCreateWebpackConfig"] =
+  async (
+    { actions, loaders, getNode, getNodesByType, pathPrefix, reporter, cache },
+    pluginOptions
+  ) => {
+    const mdxNodes = getNodesByType(`Mdx`)
+    const nodeMap: NodeMap = new Map()
+    mdxNodes.forEach(mdxNode => {
+      if (!mdxNode.parent) {
+        return
+      }
+      const fileNode = getNode(mdxNode.parent) as IFileNode
```

Comment:
I'd recommend not using type assertions (https://medium.com/swlh/why-you-should-avoid-type-assertions-in-typescript-5494e3d04dd) but instead type `mdxNodes` above or `fileNode: IFileNode` here